gecoscc/utils.py

```diff
@@ -2245,3 +2261,259 @@ def trace_inheritance(logger, db, action, obj, policy):
+    ''' Upload cookbook to chef server
+    
+    Args:
+      user(str):            user with permission for uploading cookbook to chef server
+      cookbook_path(str):   path to cookbook files
+    '''
+    logger.info("Uploading cookbook ...")
+
+    try:
+
+        settings = get_current_registry().settings
+
+        if is_cli_request():
+            has_cli_permission(os.environ['SCRIPT_CODE'], upload_cookbook.__name__)
+            user = pickle.loads(os.environ['GECOS_USER'])
+            cookbook_path = os.environ['COOKBOOK_DIR']
+
+        assert user is not None and cookbook_path is not None, _('Missing required arguments')
+        assert os.path.isdir(cookbook_path), _('Directory %s can\'t be found.') % cookbook_path
+
+        api = get_chef_api(settings, user)
+
+        admin_cert = os.sep.join([settings.get('firstboot_api.media'), user['username'], 'chef_user.pem'])
+        logger.debug("upload_cookbook: admin_cert = %s" % admin_cert)
+
+        chef_url = settings.get('chef.url') + '/organizations/default'
+        logger.debug("upload_cookbook: chef_url = %s" % chef_url)
+
+        command = 'source /opt/rh/rh-ruby24/enable; knife cookbook upload {0} -s {1} -u {2} -k {3} -o {4}'.format(settings['chef.cookbook_name'], chef_url, user['username'], admin_cert, cookbook_path)
+
+        upload_output = subprocess.check_output(command, shell=True)
+        logger.info(upload_output)
+        logger.info("Uploaded cookbook.")
+        
+    except AssertionError, msg:
+        logger.warning(msg)
+
+    except subprocess.CalledProcessError, msg:
+        logger.error(msg.cmd)
+        logger.error(msg.output)
+
+
+def chefserver_backup(username=None, backupdir=None):
+    ''' Backing up all Chef server data
+    
+    Args:
+      username(str):    user with permission in chef server
+      backupdir(str):   backup directory where hold files
+    '''
+    logger.info("Backing up Chef Server ...")
+
+    try:
+ 
+        settings = get_current_registry().settings
+
+        if is_cli_request():
+            has_cli_permission(os.environ['SCRIPT_CODE'], chefserver_backup.__name__)
+            username = pickle.loads(os.environ['GECOS_USER']).get('username',None)
+            backupdir = os.environ['BACKUP_DIR']
+
+        logger.debug("utils.py ::: chefserver_backup - username = %s" % username)
+        logger.debug("utils.py ::: chefserver_backup - backupdir = %s" % backupdir)
+
+        assert username is not None or backupdir is not None, _('Missing required arguments')
+
+        if not os.path.exists(backupdir):
+            os.mkdir(backupdir)
+
+        admin_cert = os.sep.join([settings.get('firstboot_api.media'), username, 'chef_user.pem'])
+        logger.debug("utils.py ::: chefserver_backup - admin_cert = %s" % admin_cert)
+
+        command = 'source /opt/rh/rh-ruby24/enable; knife backup export -D {0} -y -s {1} -u {2} -k {3}'.format(backupdir, settings.get('chef.url'), username, admin_cert)
```

Comment:
This call does not look easy to maintain.

Why do you call knife with these "source"?

Reply:
The previous solution is discarded and replaced by the execution of script that configures the ruby environment and executes the knife command